### DISCUSSION TITLE
Instance registry

### DISCUSSION THREAD

**Discussion by uvlad7** [ruby-rice/rice#216] [General]
It's not a some specific bug, actually, I just want to discuss instance registry problems highlighted in the docs.

> There could be a window where a Ruby object is marked for deletion but the underlying C++ object is returned back to Ruby

I feel like it's possible to reuse `ObjectSpace._id2ref` logic here, and take a look on how gc [frees `T_DATA`](https://github.com/ruby/ruby/blob/v2_7_8/gc.c#L2701). Or maybe use [weakmap](https://github.com/ruby/ruby/blob/v2_7_8/gc.c#L10423).

> Ruby instance may be freed on the C++ side

Is it possible to enable the registry only for Ruby-managed instances? Maybe use Ruby's `xmalloc` somehow?

**Comment by cfis** [comment]
The trickiest problem is when a C++ object is freed and a new one is created using the same address. I don't see way that Rice would be able to figure that out.

Any C++ created objects would not use Ruby's xmalloc call on creation - instead they will use the malloc call that they are linked against in the std c++ library.

**Reply by uvlad7** [reply]
> when a C++ object is freed

I don't get it. A C++ object must not be freed as long as Ruby object that wraps it isn't freed, right? And when Ruby object is freed, it's deleted from the registry.

If the same C++ object is returned to Ruby again, a new Ruby object will be created.

**Reply by cfis** [reply]
Yes in theory.  But either Jason or I wrote that text based on what we are seeing. 

Maybe give it a try and see what you find.
